stan/math/prim/fun/complex_schur_decompose.hpp

```diff
@@ -0,0 +1,64 @@
+#ifndef STAN_MATH_PRIM_FUN_COMPLEX_SCHUR_DECOMPOSE_HPP
+#define STAN_MATH_PRIM_FUN_COMPLEX_SCHUR_DECOMPOSE_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the unitary matrix of the complex Schur decomposition of the
+ * specified square matrix.
+ *
+ * The complex Schur decomposition of a square matrix `A` produces a
+ * complex unitary matrix `U` and a complex upper-triangular Schur
+ * form matrix `T` such that `A = U * T * inv(U)`.  Further, the
+ * unitary matrix's inverse is equal to its conjugate transpose,
+ * `inv(U) = U*`, where `U*(i, j) = conj(U(j, i))`
+ *
+ * @tparam value type of matrix
+ * @param m real matrix to decompose
+ * @return complex unitary matrix of the complex Schur decomposition of the
+ * specified matrix
+ * @see complex_schur_decompose_t
+ */
+template <typename M, require_eigen_dense_dynamic_t<M>* = nullptr>
+inline Eigen::Matrix<complex_return_t<scalar_type_t<M>>, -1, -1>
+complex_schur_decompose_u(const M& m) {
+  check_nonzero_size("complex_schur_decompose_u", "m", m);
+  check_square("complex_schur_decompose_u", "m", m);
+  using MatType = Eigen::Matrix<scalar_type_t<M>, -1, -1>;
+  // copy because ComplexSchur requires Eigen::Matrix type
+  MatType mv = m;
+  Eigen::ComplexSchur<MatType> cs(mv);
+  return cs.matrixU();
```

Comment:
It this prematurely collapsing what would otherwise be an expression template?  I think Eigen's actually returning a concrete matrix here, so hopefully no copy (they make it really hard to cut from Eigen doc):.

<tt>[Matrix](https://eigen.tuxfamily.org/dox/classEigen_1_1Matrix.html)<[ComplexScalar](https://eigen.tuxfamily.org/dox/classEigen_1_1ComplexSchur.html#a9cfc7d20234dd51933b0cb6ef427c8b4), RowsAtCompileTime, ColsAtCompileTime, Options, MaxRowsAtCompileTime, MaxColsAtCompileTime></tt>

Reply:
See above, but yes, this is forcing an expression evaluation.

Reply:
I think the return is already a concrete matrix if I'm reading the type right, so I don't think there's a better way to return this.